CIP-NFTMetadataStandard.md

```diff
@@ -0,0 +1,119 @@
+---
+CIP:
+Title: NFT Metadata Standard
+Authors: Alessandro Konrad <alessandro.konrad@live.de>, Smaug <smaug@pool.pm>
+Comments-URI:
+Status: Draft
+Type: Informational
+Created: 2021-04-08
+Post-History: https://forum.cardano.org/t/cip-nft-metadata-standard/45687 and https://www.reddit.com/r/CardanoDevelopers/comments/mkhlv8/nft_metadata_standard/
+License: CC-BY-4.0
+---
+
+## Abstract
+
+This proposal defines an NFT Metadata Standard for Native Tokens.
+
+## Motivation
+
+Tokens on Cardano are a part of the ledger. Unlike on Ethereum, where metadata can be attached to a token through a smart contract, this isn't possible on Cardano because tokens are native and Cardano uses a UTxO ledger, which makes it hard to directly attach metadata to a token.
+So the link to the metadata needs to be established differently.
+Cardano has the ability to send metadata in a transaction, that's the way we can create a link between a token and the metadata. To make the link unique, the metadata should be appended to the same transaction, where the token forge happens:
+
+> Given a token in a EUTXOma ledger, we can ask “where did this token come from?” Since tokens
+> are always created in specific forging operations, we can always trace them back through their
+> transaction graph to their origin.
+
+(Section 4.1 in the paper: https://hydra.iohk.io/build/5400786/download/1/eutxoma.pdf)
+
+## Considerations
+
+That being said, we have unique metadata link to a token and can always prove that with 100% certainty. No one else can manipulate the link except if the policy allows it to ([update mechanism](#update-metadata-link-for-a-specific-token)).
+
+## Specification
+
+This is the registered `transaction_metadatum_label` value
+
+| transaction_metadatum_label | description  |
+| --------------------------- | ------------ |
+| 721                         | NFT Metadata |
+
+### Structure
+
+The structure allows for multiple token mints, also with different policies, in a single transaction.
+
+```
+{
+  "721": {
+    "<policy_id>": {
+      "<asset_name>": {
+        "name": "<name>",
```

Comment:
> In my opinion it's crucial for many NFTs to have the metadata on-chain, even if it's just name or description

Could you elaborate on that? While I understand it may be a little bit easier for downstream applications to have metadata readily available on-chain. The cost for doing an extra query to fetch them from an off-chain registry is quite low, especially since the image itself already has to be fetched so it is not like this is introducing an extra operational cost. 

The blockchain is a ledger, not a database. It makes for a very poor choice of storage and, storing ten of thousands of NFT metadata on the blockchain does not sound like a good idea to me. Rather, the blockchain is good at recording hashes of content stored off-chain. I like the approach of associated metadata with some tokens in the minting transaction as it makes the relationship between the metadata and the token creator unambiguous. However, we should avoid storing the actual content on the chain and prefer off-chain solutions for that.

Reply:
I've following up with this thread, and I'm thinking that if you rely on IPFS to provide image, why not use it to store the metadata as well? I know the downs, but, i'm forward with the fact that we need to store as less as possible on chain, "a ledger is not a database", on the long run this is not healthy for the chain. IPFS is already largely relied upon by practically all NFTs around, and an IPFS CIDs can represent anything, a file, an image, a json, a text....

Versioning/labeling the metadata with something like 721 (not 721), is, in my honest opinion, the best approach so the community (& marketplaces) can agree on formats on how to validate the metadata that comes back from IPFS.

Reply:
> Why not use it to store the metadata as well?

This is actually one avenue we've indeed considered during the editor meeting. On the plus side, IPFS store data by hashes, so we get both addressable and verifiable content in one URI. We could very much imagine storing a JSON file on IPFS with a format defined by this CIP, and likely carrying another URI to an image hosted on IPFS.

Reply:
Let's say you buy a very expensive NFT, where the link to the image or other media file is on-chain, but the name and description is not. Now the original creator can change the name and description to something arbitrary and ridiculous. These common properties will be shown in most third parties and it could actually have an impact on the value of the NFT. You depend on the creator in some respect.

I agree that content should not be stored on-chain, especially not images or alike. But I think that name and description don't take up too much storage and are actually part of metadata. But let's assume the metadata for one specific NFT is pretty large. Instead of using the off-chain registry, I would rather use a decentralized file storage like IPFS and propose the following structure:
```
{
  "721": {
    "<policy_id>": {
      "<asset_name>": {
         "offchain": "ipfs://<hash>"
      }
    }
  }
}
```
The **`offchain`** property now contains all the properties that would have been on-chain. The IPFS link is just an example. Any other decentralized storage solution works as well. The idea is just to keep everything as decentralized as possible and don't rely on 3rd parties or the creator itself.

Reply:
That's my thoughts exactly Alessandro, thanks for the great work pushing this CIP forward btw, i'm here to help any way I can.
The on-chain metadata label can define a way for us to know what to expect from IPFS, this makes total sense to me, you can't tamper a file once it's on IPFS, it will have a new CID, so it's immutable as well (and addressable).

I understood your concern to make querying those name/descriptions easier by having it on-chain, but I don't think it makes much of a difference if you grab it from IPFS as well like you've shown above.

I think that the community is eager to flourish in terms of NFTs, and doing this off-chain, can also help on this regard, we can quickly iterate versions/labels and evolve, without having to rely on on-chain structural changes.

Reply:
Hey that's a very interesting. If I understood that right, instead of proposing more CIPs for various use cases, NFTs are minted containing the metadata structure and 3rd parties can look up the structures right? The question is who will decide which types are valid or allowed to be minted.

Regarding the off-chain approach I have no problem with IPFS. Before I was referring to the official off-chain metadata registry, which is mutable and the owner of the policy can change the metadata whenever he wants. With IPFS or other decentralized file storages you have immutability.

Reply:
Yep! Cardano Native Tokens works completely different from ERCs, so, there's an opportunity for us to approach this with freedom of thought. Cause in the end, a standard is not really required, and we cannot guarantee a standard will be followed.

So instead having one/multiple standard, we can have a type definition schema, that lays out the format for all possible NFT media/data types.

We have no control over how people mint, what people mint, and what format they'll follow, what we can have is a type definition that sets the data structure on how an audio metadata block look like, or image and so on, and this would be set by this CIP.

Then, NFT builders and marketplaces can lookup this definition via this CIP Policy ID, in order to lookup types and know how to build or what to expect from an NFT projects.